File: app/src/main/java/com/ibashkimi/lockscheduler/addeditprofile/conditions/ConditionsFragment.java
package com.ibashkimi.lockscheduler.addeditprofile.conditions;

import android.content.DialogInterface;
import android.content.Intent;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.support.transition.TransitionManager;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v4.app.FragmentTransaction;
import android.support.v7.app.AlertDialog;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;

import com.ibashkimi.lockscheduler.R;
import com.ibashkimi.lockscheduler.addeditprofile.conditions.picker.PlacePickerActivity;
import com.ibashkimi.lockscheduler.addeditprofile.conditions.picker.WifiPickerActivity;
import com.ibashkimi.lockscheduler.model.Condition;
import com.ibashkimi.lockscheduler.model.PlaceCondition;
import com.ibashkimi.lockscheduler.model.PowerCondition;
import com.ibashkimi.lockscheduler.model.TimeCondition;
import com.ibashkimi.lockscheduler.model.WifiCondition;
import com.ibashkimi.lockscheduler.model.WifiItem;
import com.ibashkimi.lockscheduler.model.prefs.AppPreferencesHelper;
import com.ibashkimi.lockscheduler.util.ConditionUtils;

import java.util.ArrayList;
import java.util.List;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.OnClick;

import static android.app.Activity.RESULT_OK;
import static com.ibashkimi.lockscheduler.addeditprofile.conditions.PlaceConditionFragment.PLACE_PICKER_REQUEST;
import static com.ibashkimi.lockscheduler.model.Condition.Type.PLACE;
import static com.ibashkimi.lockscheduler.model.Condition.Type.POWER;
import static com.ibashkimi.lockscheduler.model.Condition.Type.TIME;
import static com.ibashkimi.lockscheduler.model.Condition.Type.WIFI;

public class ConditionsFragment extends Fragment {

    public static final int REQUEST_WIFI_PICKER = 5;

    @BindView(R.id.place_layout)
    ViewGroup placeLayout;

    @BindView(R.id.place_delete)
    View placeDelete;

    @BindView(R.id.time_layout)
    ViewGroup timeLayout;

    @BindView(R.id.time_delete)
    View timeDelete;

    @BindView(R.id.wifi_layout)
    ViewGroup wifiLayout;

    @BindView(R.id.wifi_body)
    View wifiBody;

    @BindView(R.id.wifi_delete)
    View wifiDelete;

    @BindView(R.id.networks_summary)
    TextView wifiSummary;

    @BindView(R.id.power_layout)
    View powerLayout;

    @BindView(R.id.power_summary)
    TextView powerSummary;

    @BindView(R.id.power_delete)
    View powerDelete;

    private FragmentManager fragmentManager;
    private List<Condition> conditions;
    private List<WifiItem> wifiItems;
    private boolean powerWhenConnected;

    private boolean placeConditionAdded = false;
    private boolean timeConditionAdded = false;
    private boolean wifiConditionAdded = false;
    private boolean powerConditionAdded = false;

    public static ConditionsFragment newInstance() {
        return new ConditionsFragment();
    }

    public void setData(List<Condition> conditions) {
        this.conditions = conditions;
    }

    public List<Condition> assembleConditions() {
        List<Condition> conditions = new ArrayList<>(3);
        if (placeConditionAdded) {
            PlaceCondition placeCondition = getPlaceConditionFragment().assembleCondition();
            if (placeCondition != null)
                conditions.add(placeCondition);
        }
        if (timeConditionAdded) {
            TimeCondition timeCondition = getTimeConditionFragment().assembleCondition();
            if (timeCondition != null)
                conditions.add(timeCondition);
        }
        if (wifiConditionAdded) {
            if (wifiItems != null && wifiItems.size() > 0)
                conditions.add(new WifiCondition(wifiItems));
        }
        if (powerConditionAdded) {
            conditions.add(new PowerCondition(powerWhenConnected));
        }
        return conditions;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        fragmentManager = getChildFragmentManager();
    }

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        ViewGroup root = (ViewGroup) inflater.inflate(R.layout.fragment_conditions, container, false);
        ButterKnife.bind(this, root);

        if (savedInstanceState == null) {
            FragmentTransaction fragmentTransaction = fragmentManager.beginTransaction();
            if (conditions != null) {
                for (Condition condition : conditions) {
                    switch (condition.getType()) {
                        case PLACE:
                            showPlaceCondition(fragmentTransaction, (PlaceCondition) condition);
                            placeDelete.setVisibility(View.VISIBLE);
                            break;
                        case TIME:
                            showTimeCondition(fragmentTransaction, (TimeCondition) condition);
                            timeDelete.setVisibility(View.VISIBLE);
                            break;
                        case WIFI:
                            showWifiCondition((WifiCondition) condition);
                            wifiDelete.setVisibility(View.VISIBLE);
                            break;
                        case POWER:
                            showPowerCondition((PowerCondition) condition);
                            powerDelete.setVisibility(View.VISIBLE);
                            break;
                    }
                }
            }
            fragmentTransaction.commit();
        } else {
            placeConditionAdded = savedInstanceState.getBoolean("place_added");
            timeConditionAdded = savedInstanceState.getBoolean("time_added");
            wifiConditionAdded = savedInstanceState.getBoolean("wifi_added");
            powerConditionAdded = savedInstanceState.getBoolean("power_added");
            if (savedInstanceState.containsKey("wifi_items_size")) {
                int size = savedInstanceState.getInt("wifi_items_size");
                wifiItems = new ArrayList<>(size);
                for (int i = 0; i < size; i++) {
                    String ssid = savedInstanceState.getString("wifi_item_" + i, null);
                    if (ssid == null)
                        throw new IllegalStateException("Cannot restore wifi item at position " + i + ".");
                    wifiItems.add(new WifiItem(ssid));
                }
                showWifiCondition(wifiItems);
            }
            powerWhenConnected = savedInstanceState.getBoolean("when_power_connected");
            showPowerCondition(new PowerCondition(powerWhenConnected));
        }
        return root;
    }

    @Override
    public void onSaveInstanceState(Bundle outState) {
        super.onSaveInstanceState(outState);
        outState.putBoolean("place_added", placeConditionAdded);
        outState.putBoolean("time_added", timeConditionAdded);
        outState.putBoolean("wifi_added", wifiConditionAdded);
        outState.putBoolean("power_added", powerConditionAdded);
        if (wifiItems != null) {
            for (int i = 0; i < wifiItems.size(); i++) {
                outState.putString("wifi_item_" + i, wifiItems.get(i).getSsid());
            }
            outState.putInt("wifi_items_size", wifiItems.size());
        }
        outState.putBoolean("when_power_connected", powerWhenConnected);
    }

    public void showPlacePicker() {
        Intent intent = new Intent(getContext(), PlacePickerActivity.class);
        intent.putExtra("radius", 300);
        startActivityForResult(intent, PLACE_PICKER_REQUEST);
    }

    public void showPlacePicker(@NonNull PlaceCondition placeCondition) {
        Intent intent = new Intent(getContext(), PlacePickerActivity.class);
        intent.putExtra("latitude", placeCondition.getLatitude());
        intent.putExtra("longitude", placeCondition.getLongitude());
        intent.putExtra("radius", placeCondition.getRadius());
        intent.putExtra("map_type", AppPreferencesHelper.INSTANCE.getMapStyle());
        startActivityForResult(intent, PLACE_PICKER_REQUEST);
    }

    public void showWifiPicker(List<WifiItem> items) {
        Intent intent = new Intent(getContext(), WifiPickerActivity.class);

        if (items != null && items.size() > 0) {
            String[] itemReps = new String[items.size()];
            for (int i = 0; i < items.size(); i++) {
                itemReps[i] = items.get(i).getSsid();
            }
            intent.putExtra("ssids", itemReps);
        }
        startActivityForResult(intent, REQUEST_WIFI_PICKER);
    }

    @Override
    public void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == PLACE_PICKER_REQUEST) {
            if (resultCode == RESULT_OK) {
                double latitude = data.getDoubleExtra("latitude", 0);
                double longitude = data.getDoubleExtra("longitude", 0);
                float radius = data.getFloatExtra("radius", 0);
                String address = data.getStringExtra("address");

                PlaceCondition placeCondition = new PlaceCondition(latitude, longitude, (int) radius, address);

                FragmentTransaction transaction = fragmentManager.beginTransaction();
                showPlaceCondition(transaction, placeCondition);
                //http://stackoverflow.com/questions/7469082/getting-exception-illegalstateexception-can-not-perform-this-action-after-onsa
                transaction.commitAllowingStateLoss();

                placeDelete.setVisibility(View.VISIBLE);
            }
        } else if (requestCode == REQUEST_WIFI_PICKER) {
            if (resultCode == RESULT_OK) {
                String[] ssidArray = data.getStringArrayExtra("ssids");
                wifiItems = new ArrayList<>(ssidArray.length);
                for (String ssid : ssidArray) wifiItems.add(new WifiItem(ssid));
                if (wifiItems.size() > 0)
                    showWifiCondition(wifiItems);
                else
                    removeWifiCondition();
            }
        }
    }

    private void showWifiCondition(List<WifiItem> networks) {
        CharSequence[] wifiList = new CharSequence[networks.size()];
        for (int i = 0; i < wifiList.length; i++) wifiList[i] = networks.get(i).getSsid();
        wifiSummary.setText(ConditionUtils.concatenate(wifiList, ", "));
        wifiBody.setVisibility(View.VISIBLE);
        wifiDelete.setVisibility(View.VISIBLE);
    }

    @OnClick(R.id.place_layout)
    public void onPlaceLayoutClick() {
        placeConditionAdded = true;
        TransitionManager.beginDelayedTransition(placeLayout);
        showPlacePicker();
    }

    @OnClick(R.id.place_delete)
    public void onPlaceDeleteClick() {
        TransitionManager.beginDelayedTransition(placeLayout);
        placeConditionAdded = false;
        fragmentManager.beginTransaction().remove(getPlaceConditionFragment()).commit();
        placeDelete.setVisibility(View.GONE);
    }

    @OnClick(R.id.time_layout)
    public void onTimeLayoutClick() {
        timeConditionAdded = true;
        TransitionManager.beginDelayedTransition(placeLayout);
        fragmentManager
                .beginTransaction()
                .replace(R.id.time_condition_container, getTimeConditionFragment(), "time_condition")
                .commit();
        timeDelete.setVisibility(View.VISIBLE);
    }

    @OnClick(R.id.time_delete)
    public void onTimeDeleteClick() {
        TransitionManager.beginDelayedTransition(placeLayout);
        timeConditionAdded = false;
        fragmentManager.beginTransaction().remove(getTimeConditionFragment()).commit();
        timeDelete.setVisibility(View.GONE);
    }

    @OnClick({R.id.wifi_layout, R.id.wifi_body})
    public void onWifiLayoutClicked() {
        wifiConditionAdded = true;
        TransitionManager.beginDelayedTransition(placeLayout);
        showWifiPicker(wifiItems);
    }

    @OnClick(R.id.wifi_delete)
    public void removeWifiCondition() {
        TransitionManager.beginDelayedTransition(placeLayout);
        wifiConditionAdded = false;
        //fragmentManager.beginTransaction().remove(getWifiConditionFragment()).commit();
        wifiDelete.setVisibility(View.GONE);
        wifiBody.setVisibility(View.GONE);
        wifiItems = null;
    }

    @OnClick(R.id.power_layout)
    public void onPowerLayoutClicked() {
        showPowerConditionDialog();
    }

    @OnClick(R.id.power_delete)
    public void removePowerCondition() {
        powerConditionAdded = false;
        powerDelete.setVisibility(View.GONE);
        powerSummary.setVisibility(View.GONE);
    }

    private void showPlaceCondition(FragmentTransaction transaction, PlaceCondition condition) {
        placeConditionAdded = true;
        PlaceConditionFragment fragment = getPlaceConditionFragment();
        fragment.setData(condition);
        transaction.replace(R.id.place_condition_container, fragment, "place_condition");
    }

    private void showTimeCondition(FragmentTransaction transaction, TimeCondition condition) {
        timeConditionAdded = true;
        TimeConditionFragment fragment = getTimeConditionFragment();
        fragment.setData(condition);
        transaction.replace(R.id.time_condition_container, fragment, "time_condition");
    }

    private void showWifiCondition(WifiCondition condition) {
        wifiConditionAdded = true;
        wifiItems = condition.getWifiList();
        showWifiCondition(wifiItems);
    }

    private void showPowerCondition(PowerCondition condition) {
        powerSummary.setText(condition.getPowerConnected() ? R.string.power_connected : R.string.power_disconnected);
        powerSummary.setVisibility(View.VISIBLE);
        powerDelete.setVisibility(View.VISIBLE);
        powerConditionAdded = true;
    }

    private void showPowerConditionDialog() {
        String[] items = getResources().getStringArray(R.array.power_state);
        AlertDialog.Builder builder = new AlertDialog.Builder(getContext());
        builder.setTitle(R.string.power_condition_title)
                .setSingleChoiceItems(items, powerWhenConnected ? 0 : 1, new DialogInterface.OnClickListener() {
                    @Override
                    public void onClick(DialogInterface dialog, int which) {
                        powerWhenConnected = which == 0;
                        showPowerCondition(new PowerCondition(powerWhenConnected));
                        dialog.dismiss();
                    }
                })
                .create().show();
    }

    private PlaceConditionFragment getPlaceConditionFragment() {
        PlaceConditionFragment fragment = (PlaceConditionFragment) fragmentManager.findFragmentByTag("place_condition");
        if (fragment == null) {
            fragment = new PlaceConditionFragment();
        }
        return fragment;
    }

    private TimeConditionFragment getTimeConditionFragment() {
        TimeConditionFragment fragment = (TimeConditionFragment) fragmentManager.findFragmentByTag("time_condition");
        if (fragment == null) {
            fragment = new TimeConditionFragment();
        }
        return fragment;
    }
}

File: app/src/main/java/com/ibashkimi/lockscheduler/addeditprofile/conditions/ConditionsFragment.kt
package com.ibashkimi.lockscheduler.addeditprofile.conditions

import android.Manifest
import android.app.Activity.RESULT_OK
import android.content.Intent
import android.content.pm.PackageManager
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.appcompat.app.AlertDialog
import androidx.fragment.app.Fragment
import androidx.lifecycle.SavedStateViewModelFactory
import androidx.lifecycle.ViewModelProvider
import androidx.transition.TransitionManager
import com.ibashkimi.lockscheduler.App
import com.ibashkimi.lockscheduler.R
import com.ibashkimi.lockscheduler.addeditprofile.AddEditProfileViewModel
import com.ibashkimi.lockscheduler.addeditprofile.conditions.location.PlacePickerActivity
import com.ibashkimi.lockscheduler.addeditprofile.conditions.time.TimePickerActivity
import com.ibashkimi.lockscheduler.addeditprofile.conditions.wifi.SelectableWifiItem
import com.ibashkimi.lockscheduler.addeditprofile.conditions.wifi.WifiPickerActivity
import com.ibashkimi.lockscheduler.data.prefs.AppPreferencesHelper
import com.ibashkimi.lockscheduler.databinding.FragmentConditionsBinding
import com.ibashkimi.lockscheduler.extention.checkPermission
import com.ibashkimi.lockscheduler.extention.concatenate
import com.ibashkimi.lockscheduler.extention.requestPermission
import com.ibashkimi.lockscheduler.util.toDaysString
import com.ibashkimi.lockscheduler.model.condition.*
import com.ibashkimi.lockscheduler.util.Utils


class ConditionsFragment : Fragment(), View.OnClickListener {

    private lateinit var binding: FragmentConditionsBinding

    private lateinit var viewModel: AddEditProfileViewModel

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        binding = FragmentConditionsBinding.inflate(inflater, container, false)
        return binding.root
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        binding.locationLayout.setOnClickListener(this)
        binding.locationDelete.setOnClickListener(this)
        binding.powerLayout.setOnClickListener(this)
        binding.powerDelete.setOnClickListener(this)
        binding.timeLayout.setOnClickListener(this)
        binding.timeDelete.setOnClickListener(this)
        binding.wifiLayout.setOnClickListener(this)
        binding.wifiDelete.setOnClickListener(this)

        viewModel = ViewModelProvider(
            requireParentFragment(),
            SavedStateViewModelFactory(App.getInstance(), requireParentFragment())
        )
            .get(AddEditProfileViewModel::class.java)
        viewModel.getPlaceCondition()
            .observe(viewLifecycleOwner, androidx.lifecycle.Observer { placeCondition ->
                placeCondition?.let { showLocationCondition(it) } ?: removeLocationCondition()
            })
        viewModel.getPowerCondition()
            .observe(viewLifecycleOwner, androidx.lifecycle.Observer { powerCondition ->
                powerCondition?.let { showPowerCondition(it) } ?: removePowerCondition()
            })
        viewModel.getTimeCondition()
            .observe(viewLifecycleOwner, androidx.lifecycle.Observer { timeCondition ->
                timeCondition?.let { showTimeCondition(it) } ?: removeTimeCondition()
            })
        viewModel.getWifiCondition()
            .observe(viewLifecycleOwner, androidx.lifecycle.Observer { wifiCondition ->
                wifiCondition?.let { showWifiCondition(it) } ?: removeWifiCondition()
            })

    }

    override fun onClick(view: View?) {
        when (view?.id) {
            R.id.locationLayout -> showPlacePicker()
            R.id.locationDelete -> viewModel.setPlaceCondition(null)
            R.id.timeLayout -> showTimePicker()
            R.id.timeDelete -> viewModel.setTimeCondition(null)
            R.id.wifiLayout -> showWifiPicker()
            R.id.wifiDelete -> viewModel.setWifiCondition(null)
            R.id.powerLayout -> showPowerConditionDialog()
            R.id.powerDelete -> viewModel.setPowerCondition(null)
        }
    }

    private fun showPlacePicker() {
        checkPermission(Manifest.permission.ACCESS_FINE_LOCATION,
            whenGranted = {
                val placeCondition = viewModel.getPlaceCondition().value
                val intent = Intent(requireContext(), PlacePickerActivity::class.java)
                intent.putExtra("latitude", placeCondition?.latitude)
                intent.putExtra("longitude", placeCondition?.longitude)
                intent.putExtra("radius", placeCondition?.radius ?: 300)
                intent.putExtra("map_type", AppPreferencesHelper.mapStyle)
                startActivityForResult(intent, REQUEST_LOCATION_PICKER)
            },
            whenExplanationNeed = {
                AlertDialog.Builder(requireContext())
                    .setTitle(R.string.location_permission_needed)
                    .setMessage(R.string.location_permission_rationale)
                    .setNegativeButton(android.R.string.cancel) { dialogInterface, _ ->
                        dialogInterface.dismiss()
                    }
                    .setPositiveButton(android.R.string.ok) { _, _ ->
                        requestPermission(
                            Manifest.permission.ACCESS_FINE_LOCATION,
                            PERMISSION_REQUEST_LOCATION_PLACE
                        )
                    }
                    .create().show()
            },
            whenDenied = {
                requestPermission(
                    Manifest.permission.ACCESS_FINE_LOCATION,
                    PERMISSION_REQUEST_LOCATION_PLACE
                )
            }
        )
    }

    override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        when (requestCode) {
            PERMISSION_REQUEST_LOCATION_PLACE -> {
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED)
                    showPlacePicker()
            }
            PERMISSION_REQUEST_LOCATION_WIFI -> {
                if (grantResults[0] == PackageManager.PERMISSION_GRANTED)
                    showWifiPicker()
            }
        }
    }

    private fun showTimePicker() {
        val timeCondition = viewModel.getTimeCondition().value
        val intent = Intent(requireContext(), TimePickerActivity::class.java)
        if (timeCondition != null) {
            intent.putExtra("time_condition", timeCondition)
        }
        startActivityForResult(intent, REQUEST_TIME_PICKER)
    }

    private fun showWifiPicker() {
        checkPermission(Manifest.permission.ACCESS_FINE_LOCATION,
            whenGranted = {
                val wifiCondition = viewModel.getWifiCondition().value
                val intent = Intent(context, WifiPickerActivity::class.java)
                wifiCondition?.wifiList?.map {
                    SelectableWifiItem(it.ssid, it.bssid, true)
                }?.toTypedArray()?.let {
                    intent.putExtra("wifi_items", it)
                }
                startActivityForResult(intent, REQUEST_WIFI_PICKER)
            },
            whenExplanationNeed = {
                AlertDialog.Builder(requireContext())
                    .setTitle(R.string.location_permission_needed)
                    .setMessage(R.string.location_permission_rationale)
                    .setNegativeButton(android.R.string.cancel) { dialogInterface, _ ->
                        dialogInterface.dismiss()
                    }
                    .setPositiveButton(android.R.string.ok) { _, _ ->
                        requestPermission(
                            Manifest.permission.ACCESS_FINE_LOCATION,
                            PERMISSION_REQUEST_LOCATION_WIFI
                        )
                    }
                    .create().show()
            },
            whenDenied = {
                requestPermission(
                    Manifest.permission.ACCESS_FINE_LOCATION,
                    PERMISSION_REQUEST_LOCATION_WIFI
                )
            }
        )
    }

    override fun onActivityResult(requestCode: Int, resultCode: Int, data: Intent?) {
        if (requestCode == REQUEST_LOCATION_PICKER) {
            if (resultCode == RESULT_OK) {
                val latitude = data!!.getDoubleExtra("latitude", 0.0)
                val longitude = data.getDoubleExtra("longitude", 0.0)
                val radius = data.getIntExtra("radius", 0)
                val address = data.getStringExtra("address")!!

                //TransitionManager.beginDelayedTransition(locationL)
                val placeCondition = PlaceCondition(latitude, longitude, radius, address)
                viewModel.setPlaceCondition(placeCondition) //showLocationCondition(placeCondition)
            }
        } else if (requestCode == REQUEST_WIFI_PICKER) {
            if (resultCode == RESULT_OK) {
                val items: List<WifiItem> = data!!.getParcelableArrayExtra("wifi_items")?.map {
                    it as SelectableWifiItem
                }?.map {
                    WifiItem(it.SSID, it.BSSID)
                } ?: emptyList()
                viewModel.setWifiCondition(if (items.isNotEmpty()) WifiCondition(items) else null)
            }
        } else if (requestCode == REQUEST_TIME_PICKER) {
            if (resultCode == RESULT_OK) {
                viewModel.setTimeCondition(data!!.getParcelableExtra("time_condition")!!)
            }
        }
    }

    private fun showLocationCondition(condition: PlaceCondition) {
        binding.apply {
            TransitionManager.beginDelayedTransition(locationLayout)
            locationTitle.setText(R.string.location_condition_title)
            locationSummary.visibility = View.VISIBLE
            locationSummary.text =
                getString(R.string.location_summary, condition.address, condition.radius)
            locationDelete.visibility = View.VISIBLE
        }
    }

    private fun removeLocationCondition() {
        binding.apply {
            TransitionManager.beginDelayedTransition(locationLayout)
            locationDelete.visibility = View.GONE
            locationSummary.visibility = View.GONE
            locationSummary.text = null
            locationTitle.setText(R.string.location_condition_add_title)
        }
    }

    private fun showTimeCondition(condition: TimeCondition) {
        binding.apply {
            TransitionManager.beginDelayedTransition(timeLayout)
            timeSummary.visibility = View.VISIBLE
            timeSummary.text = getString(
                R.string.time_condition_summary,
                condition.daysActive.toDaysString(requireContext()),
                Utils.formatTime(condition.startTime.hour, condition.startTime.minute),
                Utils.formatTime(condition.endTime.hour, condition.endTime.minute)
            )
            timeDelete.visibility = View.VISIBLE
        }
    }

    private fun removeTimeCondition() {
        binding.apply {
            TransitionManager.beginDelayedTransition(timeLayout)
            timeDelete.visibility = View.GONE
            timeSummary.visibility = View.GONE
            timeSummary.text = null
            timeTitle.setText(R.string.time_condition_add_title)
        }
    }

    private fun showWifiCondition(condition: WifiCondition) {
        binding.apply {
            TransitionManager.beginDelayedTransition(wifiLayout)
            wifiSummary.text = condition.wifiList.map { it.ssid }.concatenate(", ")
            wifiSummary.visibility = View.VISIBLE
            wifiDelete.visibility = View.VISIBLE
            wifiTitle.setText(R.string.wifi_condition_title)
            wifiDelete.visibility = View.VISIBLE
        }
    }

    private fun removeWifiCondition() {
        binding.apply {
            TransitionManager.beginDelayedTransition(wifiLayout)
            wifiDelete.visibility = View.GONE
            wifiSummary.visibility = View.GONE
            wifiSummary.text = null
            wifiTitle.setText(R.string.wifi_condition_add_title)
        }
    }

    private fun showPowerCondition(condition: PowerCondition) {
        binding.apply {
            TransitionManager.beginDelayedTransition(powerLayout)
            powerTitle.setText(R.string.power_condition_title)
            powerSummary.setText(if (condition.powerConnected) R.string.power_connected else R.string.power_disconnected)
            powerSummary.visibility = View.VISIBLE
            powerDelete.visibility = View.VISIBLE
            powerDelete.visibility = View.VISIBLE
        }
    }

    private fun removePowerCondition() {
        binding.apply {
            TransitionManager.beginDelayedTransition(locationLayout)
            powerTitle.setText(R.string.power_condition_add_title)
            powerDelete.visibility = View.GONE
            powerSummary.visibility = View.GONE
            powerSummary.text = null
        }
    }

    private fun showPowerConditionDialog() {
        val powerCondition = viewModel.getPowerCondition().value
        val items = resources.getStringArray(R.array.power_state)
        var selectedItem = -1
        if (powerCondition != null)
            selectedItem = if (powerCondition.powerConnected) 0 else 1
        val builder = AlertDialog.Builder(requireContext())
        builder.setTitle(R.string.power_condition_title)
            .setSingleChoiceItems(items, selectedItem) { dialog, which ->
                viewModel.setPowerCondition(PowerCondition(which == 0))
                dialog.dismiss()
            }
            .create().show()
    }

    companion object {

        const val REQUEST_LOCATION_PICKER = 3
        const val REQUEST_TIME_PICKER = 4
        const val REQUEST_WIFI_PICKER = 5
        const val PERMISSION_REQUEST_LOCATION_PLACE = 1
        const val PERMISSION_REQUEST_LOCATION_WIFI = 2

        fun newInstance(): ConditionsFragment {
            return ConditionsFragment()
        }
    }
}
